Add Garden Seeds main page and go-seed (#4693)
garden-io/garden: docs/garden-seeds/languages/go-seed.md

# Go seed 🌸

{% hint style="info" %}

> If you encounter any issues or bugs 🐛 in this seed, don't hesitate to contact our [Discord Community](https://go.garden.io/discord) 🌸.

Reviewer: ```suggestion
If you encounter any issues or bugs 🐛 in this seed, don't hesitate to join our [Discord community](https://go.garden.io/discord) 🌸 for access to Garden's dedicated Community Engineers and our AI chatbot 🤖  trained on our docs.
```

{% endhint %}

> Go projects and compiled languages can run seamlessly on Garden, using our SDLC inner loop with local/remote Kubernetes environments 🍃.

Reviewer: When using acronyms always define the acronym in full with e.g. (SDLC) appended but I would also change the sentence to convey greater meaning since it reads a bit like a string of buzzwords.

```suggestion
Go projects and compiled languages can run seamlessly on Garden, with fast image rebuilds as you type, cached tests and more deployed to any number of local and remote Kubernetes environments 🍃.
```

> This seed deploys a simple Go API using Helm and Garden to template our application and seamlessly deploy it to a local Kubernetes cluster ✅.

Reviewer: Should we also let them know how easy it is for them to drop-in their own Go project?

In 5 minutes ⌛ you will deploy your own Go code into a Kubernetes cluster with a rebuild-free container using Garden's code synchronization.

> [Seed Repo](https://github.com/garden-io/go-seed)

Reviewer: ```suggestion
```

I don't think we need this since the repo is repeated down-page.

## Diagram 🖼️

> This demo uses Garden with Helm to deploy to our Kubernetes cluster; to do so, we will need to create our Helm Chart first.

Reviewer: ```suggestion
This demo scaffolds a Helm chart we'll deploy to our Kubernetes cluster.
```

Avoid communicating complexity. We're doing all of this for them.

![Quick diagram with tooling used in the project](https://ce-content.s3.fr-par.scw.cloud/golang-garden-recipe.png)

## Prerequisites

1. Install Garden CLI and get a Kubernetes cluster running locally by following our [quickstart guide](https://docs.garden.io/basics/quickstart) 🔎.

> 2. Install Python3; we need this to run/install Cookiecutter 🍪; most distros/OS already include python3, but if you don't, you can install it by following this [guide](https://docs.python-guide.org/starting/installation/#installation-guides).

Reviewer: ```suggestion
2. Ensure Python is available on your system and version 3.7 or higher by running `python --version`. MacOS and most Linux distributions will have Python already installed. Follow the official Python installation [guide](https://docs.python-guide.org/starting/installation/#installation-guides) if not. Python is required to scaffold the language seed.
```

Here I've adopted “ensure” over “install” so it's clear this step is likely not required. I've also given a small hint that both confirms Python is installed on their system and also checks their Python version is the minimum required for Cookiecutter.

3. Across this seed, we mention multiple concepts of Kubernetes and Helm, so if you are unfamiliar with any terminology, don't hesitate to contact our Community Engineering team over [Discord](https://go.garden.io/discord) 💭.

## TLDR

Get your seed running in less than five minutes by running the following commands:

{% tabs %}
{% tab title="macOS" %}

````bash
brew install garden-io/garden/garden-cli
brew install cookiecutter

> cookiecutter https://github.com/garden-io/go-seed.git

Reviewer: Why is this command repeated?

Author: There would be two sections one for Linux and one for MacOS, commands are the same though

cookiecutter https://github.com/garden-io/go-seed.git # Answers the prompts to get your brand new repository
cd ${your-project-name}
garden deploy --sync
````

`garden deploy --sync` will take you to a REPL (Read, Eval, Print, Loop) interactive terminal where you can also run different commands like `test unit` which will run the `unit test` built for this HTTP API.

If you want to get your test results with greater detail, use `get test-results unit.`

{% endtab %}

{% tab title="Linux" %}

```sh
curl -sL https://get.garden.io/install.sh | bash
python3 -m pip install --user pipx
python3 -m pipx ensurepath # Reload your terminal after this step.
pipx install cookiecutter
cookiecutter https://github.com/garden-io/go-seed.git # Answers the prompts to get your brand new repository
cd ${your-project-name}
garden deploy --sync
```

`garden deploy --sync` will take you to a REPL (Read, Eval, Print, Loop) interactive terminal where you can also run different commands like `test unit` which will run the `unit test` built for this HTTP API.

> If you want to get your test results with greater detail, use `get test-results unit.`

Reviewer: ```suggestion
If you want to get your test results with greater detail, use `get test-results unit.`

To exit the REPL at any time, type `exit`.
```

{% endtab %}

{% endtabs %}

> If you would like to understand how the configuration is working behind the scenes checkout our [go-seed repository](https://github.com/garden-io/go-seed/blob/main/README.md)

Reviewer: ```suggestion
If you would like to dive into the code used for this Garden Seed, check out our [go-seed repository](https://github.com/garden-io/go-seed/blob/main/README.md)
```

## Conclusion 🔚

Congratulations 🎉 you have reached the end of this Garden seed.

> Now you can deploy your new API and be proud of your hard work, because this project has `sync` enabled, you can simply modify your `main.go` and watch how your Go code gets rebuilt each time.

Reviewer: ```suggestion
Because this project has Code Synchronization already configured, you can simply modify your `main.go` and observe changes live on the synced code path.
```

Maybe we should add something like, “Want to modify the path or use your own Go app? Simply...” We should communicate how and where users can modify the configuration for themselves.

![Test video](https://ce-content.s3.fr-par.scw.cloud/garden-go-seed.gif)

## Next Steps ⏭️

Continue your Garden Journey. If you wish to learn more about Garden, here are a couple of resources you might find interesting:

- Join our beautiful Community on [Discord](https://go.garden.io/discord) 👋🏻.
- Give us a star in our [main repository](https://github.com/garden-io/garden) to show your support 💚.
- Read the [How Garden Works](https://docs.garden.io/basics/how-garden-works) page to understand the core concepts behind Garden.
- Read the [Helm Action configuration](https://docs.garden.io/reference/action-types/deploy/helm) document to extend your Helm configuration within your new project.
- Extend your configuration by using a [Remote K8s environment](https://docs.garden.io/kubernetes-plugins/remote-k8s).

## Contribute ➕

- Join our community in [Discord](https://go.garden.io/discord) 🎉

> - If you find any bugs 🐛 or have suggestions to improve our seeds, please don't hesitate to reach out by creating an [issue here](https://github.com/garden-io/garden/issues/new?assignees=&labels=&projects=&template=BUG_REPORT.md&title=) or by asking in our [Discord Community](https://go.garden.io/discord)🌸

Reviewer: ```suggestion
- If you find any bugs 🐛 or have suggestions to improve our seeds, please don't hesitate to reach out by creating an [issue here](https://github.com/garden-io/garden/issues/new?assignees=&labels=&projects=&template=BUG_REPORT.md&title=) or by asking in our [Discord community](https://go.garden.io/discord)🌸
```
